### ISSUE TITLE
[no-floating-promises] optionally called function that return a promise is not flagged

### ISSUE THREAD

**Issue by G-Rath** [typescript-eslint/typescript-eslint#4095]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error"]
  }
}
```

```TS
const doSomething = async (callback?: () => Promise<void>): Promise<void> => {
  callback?.();
};

doSomething();
```

```JSON
{
  "compilerOptions": {
    "target": "ES2021",
    "module": "CommonJS",
    "moduleResolution": "Node",
    "lib": ["ES2021"],
    "rootDir": "./",
    "allowJs": true,
    "checkJs": true,
    "noEmit": true,
    "strict": true,
    "resolveJsonModule": true,
    "useDefineForClassFields": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noEmitOnError": true,
    "removeComments": false,
    "esModuleInterop": true
  },
  "exclude": ["coverage", "node_modules", "lib"],
  "files": [".eslintrc.js"],
  "include": ["**/*.js", "**/*.ts"]
}
```

**Expected Result**

That both lines 2 and 5 are flagged.

**Actual Result**

Only line 5 is flagged.

**Additional Info**

This looks to be the opposite of how `@typescript-eslint/await-thenable` behaves, which lets me `await` line 2 even though the type is `Promise<void> | undefined`.

~I'm happy to attempt a PR for this.~ I'm implemented a fix for this :)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.3.0` |
| `@typescript-eslint/parser`        | `5.3.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `16.0.0` |

**Comment by JoshuaKGoldberg** [comment]
Confirmed. `?.` optional chaining broke a bunch of rules that didn't have to explicitly handle its syntax before.

> I'm happy to attempt a PR for this.

You are the best kind of person, @G-Rath. 🥰
